HPCC-34747 relax XRef CriticalBlock usage

jackdelv opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jackdelv committed:
HPCC-34747 Relax CriticalBlock in XRef scanDirectories

jakesmith reviewed (changes requested):
@jackdelv - I don't think the thread safety changes are in the right place.

As discussed offline, making the HT's thread-safe in themselves doesn't protect their callers from overlapping and finding/not-finding, and creating and adding multiple times.

I don't think you need a thread-safe impl. of the HT's, but you do need to make lookupDir and when "files" are added to (by finding and creating) thread-safe.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -1149,9 +1210,11 @@ class CNewXRefManager: public CNewXRefManagerBase
-                if (!ok||abort)
-                    return;
+                {
+                    CriticalBlock block(crit);
```
is this necessary, does it actually protect against anything? (abort might change after it leaves crit).
abort (and the stopped it is a reference to) should really be an atomic

jackdelv replied:
I'm not sure it is necessary, but was trying to keep old behavior. I can remove second abort check and CriticalBlock then change it to atomic.

jakesmith replied:
now that 'crit' is only used here, and it doesn't serve any purpose, let's remove it, and the crit.
Changing this boolean to an atomic (which has causes wider spread changes) can be done seprarely.

jackdelv replied:
Removed crit from CNewXRefManager

jakesmith reviewed (commented):
@jackdelv - please see comments.

mergeDirPerPartDirs calls set marked/setpresent, which may need protecting too. Please check.

jackdelv commented:
@jakesmith I have addressed your comments and looked at mergeDirPerPartDirs. Since mergeDirPerPartDirs happens during the listOrphans scan, it is single threaded and shouldn't need a CriticalSection. Also, should the CriticalBlock around the second !ok||abort check be removed? I'm not sure of the motivation behind the original behavior.

jakesmith reviewed (commented):
@jackdelv - looks good, minor comments only.

jakesmith reviewed (approved):
@jackdelv - looks good. Please squash.

jackdelv commented:
@jakesmith Squashed.
